src/Microsoft.DocAsCode.MarkdigEngine.Extensions/TripleColon/CodeExtension.cs

```diff
@@ -0,0 +1,287 @@
+        {
+            htmlAttributes = null;
+            renderProperties = new Dictionary<string, string>();
+            var source = string.Empty;
+            var range = string.Empty;
+            var id = string.Empty;
+            var highlight = string.Empty;
+            var language = string.Empty;
+            var interactive = string.Empty;
+
+            foreach (var attribute in attributes)
+            {
+                var name = attribute.Key;
+                var value = attribute.Value;
+                switch (name)
+                {
+                    case "source":
+                        source = value;
+                        break;
+                    case "range":
+                        range = value;
+                        break;
+                    case "id":
+                        id = value;
+                        break;
+                    case "highlight":
+                        highlight = value;
+                        break;
+                    case "language":
+                        language = value;
+                        break;
+                    case "interactive":
+                        interactive = value;
+                        break;
+                    default:
+                        logError($"Unexpected attribute \"{name}\".");
+                        return false;
+                }
+            }
+
+            if (string.IsNullOrEmpty(source))
+            {
+                logError($"source is a required attribute. Please ensure you have specified a source attribute");
+                return false;
+            }
+
+            if(string.IsNullOrEmpty(language))
+            {
+                language = InferLanguageFromFile(source, logError);
+            }
+
+            htmlAttributes = new HtmlAttributes();
+            htmlAttributes.AddProperty("class", $"lang-{language}");
+            if (!string.IsNullOrEmpty(interactive))
+            {
+                htmlAttributes.AddProperty("data-interactive", language);
+                htmlAttributes.AddProperty("data-interactive-mode", interactive);
+            }
+            if (!string.IsNullOrEmpty(highlight)) htmlAttributes.AddProperty("highlight-lines", highlight);
+
+            RenderDelegate = (renderer, obj) =>
+            {
+                var currentId = string.Empty;
+                var currentRange = string.Empty;
+                var currentSource = string.Empty;
+                obj.Attributes.TryGetValue("id", out currentId); //it's okay if this is null
+                obj.Attributes.TryGetValue("range", out currentRange); //it's okay if this is null
+                obj.Attributes.TryGetValue("source", out currentSource); //source has already been checked above
+                var (code, codePath) = _context.ReadFile(currentSource, InclusionContext.File, null);
```

Comment:
Pass `obj` to `ReadFile` to produce the correct source info for warnings raised while resolving external files.
```suggestion
                var (code, codePath) = _context.ReadFile(currentSource, InclusionContext.File, obj);
```